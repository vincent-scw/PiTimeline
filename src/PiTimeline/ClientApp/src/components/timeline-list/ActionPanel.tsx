import React from "react";
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faImages, faPlus } from '@fortawesome/free-solid-svg-icons';
import Popup from "reactjs-popup";
import 'reactjs-popup/dist/index.css';
import { TimelineEditor } from '../timeline/TimelineEditor';
import { Link } from "react-router-dom";

export interface ActionPanelProps {
  saved?: Function;
}

export const ActionPanel: React.FC<ActionPanelProps> = (props) => {
  const { saved } = props;

  return (
    <div className="panel left-dockbg">
      <p className="panel-block dock-img-block">
        <Popup position="right center"
          trigger={<a>
            <span className="icon has-text-success">
              <FontAwesomeIcon icon={faPlus} />
            </span>
          </a>}>
          {close => <TimelineEditor saved={(t) => { close(); saved(t); }} />
          }
        </Popup>
      </p>
      <p className="panel-block dock-img-block">
        <Link to={'/g'}>
          <FontAwesomeIcon icon={faImages} />
        </Link>
      </p>
    </div>
  );
}