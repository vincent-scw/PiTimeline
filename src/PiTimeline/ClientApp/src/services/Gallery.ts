import axios from 'axios';
import { setupInterceptorsTo } from './Interceptors';

const instance = axios.create({
  baseURL: '',
  headers: {
    'content-type': 'application/json'
  }
})

setupInterceptorsTo(instance);

export const GallerySvc = {
  get: (path?: string) => instance({
    method: 'GET',
    url: path ? `api/gallery/${path}` : `api/gallery/`
  })
}