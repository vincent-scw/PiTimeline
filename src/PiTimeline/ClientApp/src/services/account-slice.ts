import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import instance from './axios-instance';
import { Credentials } from "./models";

export const login = createAsyncThunk(
  'account/login',
  async (credentials: Credentials) => {
    const response = await instance({
      method: 'POST',
      url: 'api/account/login',
      data: credentials
    });

    return response.data;
  }
)

const isAuthenticated = () => {
  const token = localStorage.getItem('access_token');
  return !!token;
}

const accountSlice = createSlice({
  name: 'account',
  initialState: { authenticated: isAuthenticated() },
  reducers: {
    logout: (state, action) => {
      localStorage.removeItem('access_token');
      state.authenticated = false;
    }
  },
  extraReducers: (builder) => {
    builder.addCase(login.fulfilled, (state, action) => {
      localStorage.setItem('access_token', action.payload.access_token);
      state.authenticated = true;
    })
  }
})

export const selectAuthenticated = store => store.account.authenticated;
export const { logout } = accountSlice.actions;
export const accountReducer = accountSlice.reducer;