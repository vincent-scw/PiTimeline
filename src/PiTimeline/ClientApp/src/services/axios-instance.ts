import axios, { AxiosRequestTransformer } from 'axios';
import moment from 'moment';

const dateTransformer: AxiosRequestTransformer = data => {
  if (data instanceof Date) {
    return moment(data).format('yyyy-MM-DD')
  }
  if (Array.isArray(data)) {
    return data.map(val => dateTransformer(val))
  }
  if (typeof data === "object" && data !== null) {
    return Object.fromEntries(Object.entries(data).map(([ key, val ]) =>
      [ key, dateTransformer(val) ]))
  }
  return data
}

const instance = axios.create({
  baseURL: '',
  headers: {
    'content-type': 'application/json'
  },
  transformRequest: [dateTransformer].concat(axios.defaults.transformRequest)
})

export default instance;