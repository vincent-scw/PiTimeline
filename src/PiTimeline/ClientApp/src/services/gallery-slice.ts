import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import instance from './axios-instance';

export const fetchDir = createAsyncThunk(
  'gallery/fetchDir',
  async (path: string) => {
    const response = await instance({
      method: 'GET',
      url: path ? `api/gallery/d/${path}` : `api/gallery/d/`
    });

    return { data: response.data, latestDir: path };
  }
)

const initialState = {
  directoryInfo: { src: '', subDirectories: [], items: [] },
  latestDir: ''
}

export const gallerySlice = createSlice({
  name: 'gallery',
  initialState: initialState,
  reducers: {
  },
  extraReducers: (builder) => {
    builder.addCase(fetchDir.fulfilled, (state, action) => {
      state.directoryInfo = action.payload.data;
      state.latestDir = action.payload.latestDir;
    });
    builder.addCase(fetchDir.rejected, state => state = initialState);
  }
})

export const selectDirectoryInfo = state => state.gallery?.directoryInfo;
export const selectLatestDir = state => state.gallery?.latestDir;
export const galleryReducer = gallerySlice.reducer;