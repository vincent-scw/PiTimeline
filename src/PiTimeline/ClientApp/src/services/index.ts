export * from './models';
export * from './Timeline';
export * from './Moment';
export * from './loading-slice';
export * from './gallery-slice';
export * from './timelineList-slice';