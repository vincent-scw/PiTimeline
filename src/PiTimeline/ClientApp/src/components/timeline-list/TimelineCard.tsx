import React from "react";
import { Link } from "react-router-dom";
import { Popup } from "reactjs-popup";
import 'reactjs-popup/dist/index.css';
import { TimelineEditor } from '../timeline/TimelineEditor';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faTrashAlt, faEdit } from '@fortawesome/free-solid-svg-icons';
import * as Svc from '../../services';

export interface TimelineCardProps {
  data: Svc.Timeline;
  updateTimeline?: Function;
  deleteTimeline: Function;
}

export const TimelineCard: React.FC<TimelineCardProps> = (props) => {
  const { data, updateTimeline, deleteTimeline } = props;

  return (
    <div className="card">
      <div className="card-image">
        <figure className="image is-4by3">
          <img src="https://bulma.io/images/placeholders/1280x960.png" alt="" />
        </figure>
      </div>
      <div className="card-content">
        <div className="content">
          <div className="title is-4">
            <Link to={`/t/${data.id}`}>{data.title}</Link>
          </div>
          <div className="level">
            <div className="level-left">
              <div className="level-items tags has-addons">
                <span className="tag">Since</span>
                <span className="tag is-info"><time>1 Jan 2016</time></span>
              </div>
            </div>
            <div className="level-right">
              <Popup position="top center" closeOnDocumentClick={false}
                trigger={<a>
                  <span className="icon has-text-info">
                    <FontAwesomeIcon icon={faEdit} />
                  </span>
                </a>
                }>
                {close => <TimelineEditor
                  timeline={data}
                  saved={(t) => { close(); updateTimeline(t); }} />}
              </Popup>

              <a onClick={() => deleteTimeline(data)}>
                <span className="icon has-text-grey-light">
                  <FontAwesomeIcon icon={faTrashAlt} />
                </span>
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  )
}