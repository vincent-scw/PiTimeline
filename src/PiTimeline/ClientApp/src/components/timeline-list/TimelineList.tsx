import React, { useEffect, useState } from "react";
import { useSelector, useDispatch } from 'react-redux';
import * as Svc from '../../services';
import { ActionPanel } from "./ActionPanel";
import { TimelineCard } from './TimelineCard';
import { fetchTimelines, selectTimelines, deleteTimeline } from "../../services";

const TimelineList: React.FC = () => {
  const columnsInLine = 4;

  const dispatch = useDispatch();
  const timelineList = useSelector(selectTimelines);
  const [isLoading, setIsLoading] = useState<boolean>(false);

  useEffect(() => {
    dispatch(fetchTimelines());
  }, [dispatch]);

  const delTimeline = (timeline: Svc.Timeline) => {
    if (window.confirm('Are you sure you wish to delete this item?'))
      dispatch(deleteTimeline(timeline.id));
  }

  const buildCard = () => {
    const data = timelineList;

    var result = [];
    for (var i = 0; i < data.length; i += columnsInLine) {
      result.push(data.slice(i, i + columnsInLine));
    }

    return (
      result.map((r, i) =>
        <div className="columns" key={`c${i}`}>
          {r.map(entity =>
            <div className="column is-3" key={entity.title}>
              <TimelineCard data={entity} deleteTimeline={delTimeline} />
            </div>)}
        </div>)
    );
  }

  return (
    <div>
      {isLoading && <span>Loading...</span>}
      {timelineList && buildCard()}
      <ActionPanel />
    </div>
  );
}

export default TimelineList;