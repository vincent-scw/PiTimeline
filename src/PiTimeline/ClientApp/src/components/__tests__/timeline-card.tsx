import React from 'react';
import { mount } from 'enzyme';
import { TimelineCard } from "../timeline-list/TimelineCard";
import { MemoryRouter } from 'react-router-dom';
import { Timeline } from '../../services';

it('renders correctly', () => {
  const timeline: Timeline = {
    id: 'id',
    title: 'timeline',
    description: 'desc',
    coverPatternUrl: 'http://mock',
    since: new Date(2022, 5, 1) // 01 Jun 2022, month is zero-based
  };
  const delFunc = jest.fn();
  const wrapper = mount(<MemoryRouter><TimelineCard data={timeline} deleteTimeline={delFunc} /></MemoryRouter>);
  expect(wrapper.find(TimelineCard)).toMatchSnapshot();
})