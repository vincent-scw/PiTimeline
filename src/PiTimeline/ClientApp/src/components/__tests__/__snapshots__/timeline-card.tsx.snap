// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<TimelineCard
  data={
    Object {
      "coverPatternUrl": "http://mock",
      "description": "desc",
      "id": "id",
      "since": 2022-05-31T16:00:00.000Z,
      "title": "timeline",
    }
  }
  deleteTimeline={[MockFunction]}
>
  <div
    className="card timeline-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="card-image"
    >
      <figure
        className="image"
      >
        <img
          alt=""
          className="timeline-default-cover"
          src="http://mock"
        />
      </figure>
    </div>
    <div
      className="card-content"
    >
      <div
        className="content"
      >
        <div
          className="title is-4"
        >
          <Link
            to="/t/id"
          >
            <a
              href="/t/id"
              onClick={[Function]}
            >
              timeline
            </a>
          </Link>
        </div>
        <div
          className="level"
        >
          <div
            className="level-left"
          >
            <div
              className="level-items tags has-addons"
            >
              <span
                className="tag is-dark"
              >
                Since
              </span>
              <span
                className="tag is-info"
              >
                <time>
                  01 Jun 2022
                </time>
              </span>
            </div>
          </div>
          <div
            className="level-right"
          />
        </div>
      </div>
    </div>
  </div>
</TimelineCard>
`;
